module/executiondatasync/optimistic_sync/pipeline/pipeline.go

```diff
@@ -0,0 +1,404 @@
+				}
+				return nil
+			case optimistic_sync.StateComplete:
+				return nil // terminate
+			default:
+				return fmt.Errorf("invalid pipeline state: %s", currentState)
+			}
+		}
+	}
+}
+
+// onStartProcessing performs the initial state transitions depending on the parent state:
+// - Pending -> Processing
+// - Pending -> Abandoned
+// No errors are expected during normal operations.
+func (p *Pipeline) onStartProcessing() error {
+	switch p.parentState() {
+	case optimistic_sync.StateProcessing, optimistic_sync.StateWaitingPersist, optimistic_sync.StateComplete:
+		err := p.transitionTo(optimistic_sync.StateProcessing)
+		if err != nil {
+			return err
+		}
+		p.worker.Submit(p.performDownload)
+	case optimistic_sync.StatePending:
+		return nil
+	case optimistic_sync.StateAbandoned:
+		return p.transitionTo(optimistic_sync.StateAbandoned)
+	default:
+		// it's unexpected for the parent to be in any other state. this most likely indicates there's a bug
+		return fmt.Errorf("unexpected parent state: %s", p.parentState())
+	}
+	return nil
+}
+
+// onProcessing performs the state transitions when the pipeline is in the Processing state.
+// When data has been successfully indexed, we can transition to StateWaitingPersist.
+// No errors are expected during normal operations.
+func (p *Pipeline) onProcessing() error {
+	if p.isIndexed.Load() {
+		return p.transitionTo(optimistic_sync.StateWaitingPersist)
+	}
+	return nil
+}
+
+// onPersistChanges performs the state transitions when the pipeline is in the WaitingPersist state.
+// When the execution result has been sealed and the parent has already transitioned to StateComplete then
+// we can persist the data and transition to StateComplete.
+// No errors are expected during normal operations.
+func (p *Pipeline) onPersistChanges() error {
+	if p.isSealed.Load() && p.parentState() == optimistic_sync.StateComplete {
+		if err := p.core.Persist(); err != nil {
+			return fmt.Errorf("could not persist pending changes: %w", err)
+		}
+		return p.transitionTo(optimistic_sync.StateComplete)
+	} else {
+		return nil
+	}
+}
+
+// checkAbandoned returns true if the pipeline or its parent are abandoned.
+func (p *Pipeline) checkAbandoned() bool {
+	if p.isAbandoned.Load() {
+		return true
+	}
+	if p.parentState() == optimistic_sync.StateAbandoned {
+		return true
+	}
+	return p.GetState() == optimistic_sync.StateAbandoned
+}
+
+// GetState returns the current state of the pipeline.
+func (p *Pipeline) GetState() optimistic_sync.State {
+	return optimistic_sync.State(p.state.Load())
+}
+
+// parentState returns the last cached parent state of the pipeline.
+func (p *Pipeline) parentState() optimistic_sync.State {
+	return optimistic_sync.State(p.parentStateCache.Load())
+}
+
+// SetSealed marks the execution result as sealed.
+// This will cause the pipeline to eventually transition to the StateComplete state when the parent finishes processing.
+func (p *Pipeline) SetSealed() {
```

Comment:
currently, the implementation allows to call `SetSealed` and then `Abandon` and vice versa. This should ideally not be possible.